fyne-io/fyne
# Raster with pixels static img

## Internal optimisation of canvas rasterizer objects.

Store the "current image" inside the Raster struct, and on each paint iteration, re-use the existing image.

Only create a new image if the rectangular bounds change.

## Bad Bits:

- Adds some ugly to the calls to Generator, because of the need to pass a reference to a *Raster.  So it ends up with headscratching code that looks like this : 

`img := dest.Generator(6, 6, dest)` . (canvas/raster_test.go).  Which looks very non-idiomatic go, but cant be helped really.


## Good Bits:

- Changes do not affect existing API.  So examples build and run without change.

## Test Results

- Add some `memdebug.Print()` calls to the rendering loop in examples/life. 

- Run examples/life with the existing code. Observe that large chunks of memory allocation happen each paint iteration.  Make the screen large == problem gets worse.  

- Observe that after a while (5 - 10 minutes depending on how bad your machine is),  GC hits start to backlog the rasterization loop, with a result that memory usage climbs, and number of goroutines starts to spike.   Machine quickly becomes unresponsive.

- Re-run the tests using the patch from this PR.  Observe that there is hardly any memory allocation / GC activity now.  Running the test for 10-15 minutes shows no slowdown.

- Run the tests again with the new code + memdebug profiling.   Lets be game this time and fire up a dozen instances of life / pong.   Resize some windows occassionally.  Go get some lunch and check that in an hour, its all running smooth still.
